### PR TITLE
Copy pmap arguments to device in parallel.

### PR DESCRIPTION
**Pull request by hawkinsp**
Adds a new device_put_many method and uses it in pmap.

### CODE REVIEW

File: jax/interpreters/xla.py [hunk 1]
```diff
@@ -97,6 +97,29 @@ def device_put(x, device_num=0):
   else:
     return xb.device_put(x, device_num)  # round-trips tuple elements
 
+def device_put_many(xs_and_devices):
```
Comment (mattjj) [inline jax/interpreters/xla.py]: I know this is rich coming from me, but how about a docstring? I realize that the code is getting a bit trickier.

Here's a first attempt:

```python
  """Place multiple Python values on multiple devices in parallel.

  This is a wrapper around jax.lib.xla_bridge.device_put_many to handle
  additional Python array types, namely DeviceArray (which is already backed by
  device memory, though may be on the wrong device) and DeviceConstant. In
  particular, it avoids transferring data already placed on the correct device,
  and handles instantiating DeviceConstants.

  Args:
    xs_and_devices: a sequence of (pyval, device_num) pairs where pyval is an
      ndarray, DeviceArray, DeviceConstant, or JaxTuple and device_num is an int
      representing the physical device number on which pyval should be placed.

  Returns:
    A sequence of buffers representing the inputs placed on the corresponding
    device numbers.
  """
```

File: jax/interpreters/xla.py [hunk 2]
```diff
@@ -97,6 +97,29 @@ def device_put(x, device_num=0):
...
+      else:
+        transfer_indices.append(i)
+        transfers.append((x.device_buffer.to_py(), device_num))
+    elif isinstance(x, DeviceConstant):
+      outputs[i] = instantiate_device_constant(x, device_num=device_num)
+    else:
+      transfer_indices.append(i)
+      transfers.append((x, device_num))
+
+  transfer_results = xb.device_put_many(transfers)
```
Comment (mattjj) [inline jax/interpreters/xla.py]: Nice! Is there a TODO to add here or a few lines up about avoiding round-tripping when `x` is a `DeviceArray` but on the wrong device?

**Review by mattjj — APPROVED** [review]
Very clear!

File: jax/interpreters/xla.py (hunk 2 above)
Comment (hawkinsp) [reply jax/interpreters/xla.py]: All done.

Comment (hawkinsp) [reply jax/interpreters/xla.py]: Done.

**Pull request merged**